### PR TITLE
[Merged by Bors] - chore: remove unnecessary uses of CategoryTheory.Equivalence.mk

### PR DESCRIPTION
**Pull request by joelriou**
An equivalence of categories consists of a functor, an inverse, and unit and counit isomorphisms which satisfy the same compatibilities as for adjoint functors. However, certain equivalences are constructed via the constructor `Equivalence.mk` which does not assume these compatibilities: by changing the unit isomorphism, it is possible to satisfy the compatibility. When only the existence of an equivalence matters, we may certainly use `Equivalence.mk` (e.g. inside proofs). However, the more we develop the library, the more we may need to unfold the definitions of unit/counit isomorphisms (as it happened in #12728). Then, I would suggest that we refrain from using `Equivalence.mk` as much as possible. In this PR, most uses of `Equivalence.mk` are removed: in most cases, the compatibilities satisfied by adjunctions were easy to prove.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Category/PartialFun.lean [hunk 1]
```diff
@@ -145,13 +146,19 @@ noncomputable def partialFunEquivPointed : PartialFun.{u} ≌ Pointed :=
...
+              exact ⟨b, ⟨ha, h.symm⟩, rfl⟩
+  counitIso :=
     NatIso.ofComponents
       (fun X ↦ Pointed.Iso.mk (by classical exact Equiv.optionSubtypeNe X.point) (by rfl))
       fun {X Y} f ↦ Pointed.Hom.ext <| funext fun a ↦ by
         obtain _ | ⟨a, ha⟩ := a
         · exact f.map_point.symm
         simp_all [Option.casesOn'_eq_elim, Part.elim_toOption]
+  functor_unitIso_comp X := by
+    ext (_|x)
```
Comment (b-mehta) [inline Mathlib/CategoryTheory/Category/PartialFun.lean:158]: I'm slightly surprised aesop_cat can't do this?

File: Mathlib/CategoryTheory/Limits/Fubini.lean [hunk 2]
```diff
@@ -507,28 +507,19 @@ noncomputable def limitCurrySwapCompLimIsoLimitCurryCompLim :
...
-  erw [NatTrans.id_app]
-  -- Why can't `simp` do this?
+  dsimp [limitCurrySwapCompLimIsoLimitCurryCompLim, Equivalence.counit]
+  rw [Category.assoc, Category.assoc, limitIsoLimitCurryCompLim_hom_π_π,
+    HasLimit.isoOfEquivalence_hom_π]
   dsimp
-  -- Porting note: the original proof only had `simp`.
-  -- However, now `CategoryTheory.Bifunctor.map_id` does not get used by `simp`
-  rw [CategoryTheory.Bifunctor.map_id]
+  erw [G.map_id]
```
Comment (b-mehta) [inline Mathlib/CategoryTheory/Limits/Fubini.lean:514]: Is there a way to avoid the erw here?

Comment (joelriou) [reply Mathlib/CategoryTheory/Limits/Fubini.lean]: I have managed to remove this one.

File: Mathlib/CategoryTheory/Category/PartialFun.lean (hunk 1 above)
Comment (joelriou) [reply Mathlib/CategoryTheory/Category/PartialFun.lean:158]: No, it cannot.

**Comment by dagurtomas** [comment]
maintainer merge

**Comment by mattrobball** [comment]
!bench

File: Mathlib/CategoryTheory/Category/PartialFun.lean (hunk 1 above)
Comment (mattrobball) [inline Mathlib/CategoryTheory/Category/PartialFun.lean:158]:
```suggestion
    ext (_ | x)
```

**Comment by mattrobball** [comment]
Very nice.

**Comment by mattrobball** [comment]
bors delegate+

Thanks!

File: Mathlib/CategoryTheory/Sums/Associator.lean [hunk 3]
```diff
@@ -112,14 +112,14 @@ theorem inverseAssociator_map_inr_inr {X Y : E} (f : inr (inr X) ⟶ inr (inr Y)
...
-      (by rintro ((_|_)|_) ((_|_)|_) f <;> first | cases f | aesop_cat))
-    (NatIso.ofComponents (fun X => eqToIso
-      (by rcases X with (_|(_|_)) <;> rfl)) -- Porting note: aesop_cat fails
-      (by rintro (_|(_|_)) (_|(_|_)) f <;> first | cases f | aesop_cat))
+@[simps functor inverse]
+def associativity : (C ⊕ D) ⊕ E ≌ C ⊕ (D ⊕ E) where
+  functor := associator C D E
+  inverse := inverseAssociator C D E
+  unitIso := NatIso.ofComponents (by rintro ((_|_)|_) <;> exact Iso.refl _) (by
+    rintro ((_|_)|_) ((_|_)|_) f <;> first | cases f | aesop_cat)
```
[the remark this answers is not in the document]
Comment (joelriou) [reply Mathlib/CategoryTheory/Sums/Associator.lean]: I have fixed this one. (There are many occurrences of `_|_` in other places in mathlib.)

[the remark this answers is not in the document]
Comment (mattrobball) [reply Mathlib/CategoryTheory/Sums/Associator.lean]: Unfortunate, but not a responsibility here.

**Comment by joelriou** [comment]
> Very nice.

Actually, I believe that at some point, some proofs of triangle compatibilities were removed because of performances issues. The situation looks better now :-)

**Comment by mattrobball** [comment]
When it passes CI, please merge

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**